salt/states/x509.py

```diff
@@ -362,10 +366,112 @@ def csr_managed(name,
+        except KeyError:
+            pass
+
+    diff = []
+    for k, v in six.iteritems(required_cert_info):
+        try:
+            if v != cert_info[k]:
+                diff.append(k)
+        except KeyError:
+            diff.append(k)
+
+    return len(diff) == 0, diff
+
+
+def _certificate_days_remaining(cert_info):
+    '''
+    Get the days remaining on a certificate, defaulting to 0 if an error occurs.
+    '''
+    try:
+        expiry = cert_info['Not After']
+        return (
+            datetime.datetime.strptime(expiry, '%Y-%m-%d %H:%M:%S') -
+            datetime.datetime.now()).days
+    except KeyError:
+        return 0
+
+
+def _certificate_is_valid(name, days_remaining, append_certs, **cert_spec):
+    '''
+    Return True if the given certificate file exists, is a certificate, matches the given specification, and has the required days remaining.
+
+    If False, also provide a message explaining why.
+    '''
+    if not os.path.isfile(name):
+        return False, '{0} does not exist'.format(name), {}
+
+    try:
+        cert_info = __salt__['x509.read_certificate'](certificate=name)
+        required_cert_info = __salt__['x509.create_certificate'](testrun=True, **cert_spec)
+        if not isinstance(required_cert_info, dict):
+            raise salt.exceptions.SaltInvocationError(
+                'Unable to create new certificate: x509 module error: {0}'.format(required_cert_info))
+
+        try:
+            issuer_public_key = required_cert_info['Issuer Public Key']
+            # Verify the certificate has been signed by the ca_server or private_signing_key
+            if not __salt__['x509.verify_signature'](name, issuer_public_key):
+                errmsg = 'Certificate is not signed by private_signing_key' if 'signing_private_key' in cert_spec else 'Certificate is not signed by the requested issuer'
+                return False, errmsg, cert_info
+        except KeyError:
+            return False, 'New certificate does not include signing information', cert_info
+
+        matches, diff = _certificate_info_matches(
+            cert_info,
+            required_cert_info,
+            check_serial='serial_number' in cert_spec
+        )
+        if not matches:
+            return False, 'Certificate properties are different: {0}'.format(', '.join(diff)), cert_info
+
+        actual_days_remaining = _certificate_days_remaining(cert_info)
+        if days_remaining != 0 and actual_days_remaining < days_remaining:
+            return False, 'Certificate needs renewal: {0} days remaining but it needs to be at least {1}'.format(actual_days_remaining, days_remaining), cert_info
+
+        return True, '', cert_info
+    except salt.exceptions.SaltInvocationError as e:
+        return False, '{0} is not a valid certificate: {1}'.format(name, str(e)), {}
 
 def certificate_managed(name,
                         days_remaining=90,
-                        managed_private_key=None,
```

Comment:
This would require users to update their states with no warning. Going forward we want to try to ensure we do not require users to update their state files, unless there is good reason. Is it possible to keep this same syntax for backward compatibility?

Reply:
`managed_private_key` is currently broken and has not worked correctly for a while.

Reply:
we still cannot rip out a kwarg entirely without warning, and I assume it was working at one point, so I still think we need to ensure we keep backwards compatibility.

Reply:
It hasn't worked since at least v2016.11.2 according to https://github.com/saltstack/salt/issues/39608#issuecomment-342346894, I highly doubt anyone is actually using the kwarg successfully. I applaud the intent to keep backwards compatibility, but the only thing this rips out is a bug.

How about something like this:

```python
    if 'managed_private_key' in kwargs:
        raise salt.exceptions.SaltInvocationError(
            'managed_private_key is no longer supported by x509.certificate_managed, use a separate x509.private_key_managed call instead.')
```

Reply:
can we not make a call to private_key_managed() if its set in kwargs? A user could be upgrading from v2016.11.2 so if we are going to rip it out I still think it needs to be properly deprecated: https://docs.saltstack.com/en/latest/topics/development/deprecations.html

Reply:
If it was that simple then it wouldn't be broken.
I don't understand the rigid adherence to a deprecation schedule when it was already "ripped out" years ago. Even if you mark it deprecated now instead of removing it, it still won't work.

IMO it's bad design in the first place and should never have been added. Let each state manage its own thing, instead of also having one of them do both.

Reply:
The rigid adherence is to ensure we do not rip out features from users and force them to update their states, unless there is good reason to do so.  I don't view a regression/bug as ripping out that feature, but something that should be fixed. From the comments I've read the reason to remove this feature is because it was not working since 2017.7 and to make the function more simple. I think the right approach would be to fix, but maybe i'm not understanding why its difficult to fix, since it was working at one time?

Reply:
I appreciate the perspective from both of you. 

## Scenario

Let's actually consider what a user on 2016.11.2 will face when upgrading to 2017.7 or any other future version:

* Let's assume that `managed_private_key` did work in 2016.11.2, even though [this comment](https://github.com/saltstack/salt/issues/39608#issuecomment-342346894) doesn't actually guarantee that (_"at **least** v2016.11.2"_).
* Let's also assume that the only thing they have to worry about breaking is `x509.certificate_managed`, forgetting about the myriad of bugs and breaking changes that 2017.7 introduced that will **force the user to update their states**.

When upgrading to 2017.7 and using `x509.certificate_managed` with `managed_private_key`, the user will see an error. Nothing useful, just a stack trace dumped to the highstate output. There will be nothing in the changelog for 2017.7 telling them about this breaking change, no deprecation warnings will have been in place for the two versions before. They'll just have to search online, and probably stumble on one of the issues this PR is aiming to fix. **They will have to update their states**.

Same if they try and upgrade to any other version, including the current 3000.1. A Python stack trace with no obvious way to fix it.

## This PR improves things, it doesn't break anything

This pull request does not change any of that. `managed_private_key` is still broken in all previous versions since 2016.11, and none of the versions offer a way to fix it without the **user updating their states**.

What it does do is give a **useful error** instead of a stacktrace dumped to the highstate output. It tells the user upgrading from 2016.11 how to fix it (by updating their states) in a clear and efficient way. No searching online, no trawling through Github issues.

I'm going to guess that Salt's deprecation policy is for two reasons:

* To warn users about upcoming changes and help them to fix their states
* To reduce the support burden on the your team @Ch3LL.

In my view, showing a _clear_ and _actionable_ error message is 100% better than what is there currently.

> I think the right approach would be to fix

I'd agree if this problem had been identified in 2016.11 and fixed for 2017.7. At this point I think that ship has sailed. I can't imagine employees at SaltStack, nor people in the open-source community, will have the time to fix the bug for a version that is now beyond life support:

![Screen Shot 2020-04-08 at 19 29 11](https://user-images.githubusercontent.com/578458/78820401-a0221200-79cf-11ea-9500-4a9c954f24ce.png)

Since this PR is targeted against `master`, I believe this means it will go into a `3000.x` bugfix release, **not** 2017.7 or any other releases leading up to today. `managed_private_key` is irreversibly broken in those versions. If any users really need to be on those older versions with working x509, Salt can be patched quite easily to use this updated state (I myself am running this on 2019.2.3).

@Ch3LL I urge you to reconsider. I will not be adding `managed_private_key` back in my branch. You're welcome to use my code to patch it yourself, but I won't.

Reply:
Given that there's [apparently a workaround](https://github.com/saltstack/salt/issues/39608#issuecomment-285440745), that at *least* [two](https://github.com/saltstack/salt/issues/39608#issuecomment-285467363) [people](https://github.com/saltstack/salt/issues/39608#issuecomment-342346894) used with success, **I definitely agree with @Ch3LL here**.

It looks like it is possible to create a set of salt states with `managed_private_key` that works out of the box. Removing this without deprecation will break anyone who is currently using that workaround.

It would be different if there was no workaround and every single scenario involving `managed_private_key` failed to manage the key. Or am I misreading that issue?

Reply:
> Given that there's apparently a workaround

@waynew, that is literally the exact workaround this PR is proposing in the error message.

Reply:
@waynew You misunderstand. The state `certificate_managed` advertises a parameter (that doesn't work called `managed_private_key`). The workaround is to use the separate `private_key_managed` state instead, which does exactly what the broken parameter is supposed to do.

Removing the broken parameter will have no effect on anyone.

Reply:
@waynew this workaround is working _by removing_ `managed_private_key` from `x509.certificate_managed` and creating separate state using `x509.private_key_managed`.

In my point of view there is a bug in current documentation, provided example with `managed_private_key` will not work. And this PR will fix this bug.

Reply:
There seems to be some confusion here.

### Before

This is the usage we are removing. It doesn't work.

```yaml                                                         
/etc/pki/ca.crt:                                                
  x509.certificate_managed:                                     
    - signing_private_key: /etc/pki/ca.key                      
    - CN: ca.example.com                                                                             
    - basicConstraints: "critical CA:true"     
    - managed_private_key:
         name: /etc/pki/ca.key
         bits: 4096
         backup: True 
```

### After

This is the workaround and what this PR tells the user to do.

```yaml
/etc/pki/ca.key:                                                
  x509.private_key_managed:                                     
    - bits: 4096                                                
    - backup: True                                                                                     
                                                                
/etc/pki/ca.crt:                                                
  x509.certificate_managed:                                     
    - signing_private_key: /etc/pki/ca.key                      
    - CN: ca.example.com                                                                             
    - basicConstraints: "critical CA:true"                                                                 
    - require:                                                                                     
      - x509: /etc/pki/ca.key    
```

Reply:
This seems to work for me:

```
/tmp/sample.key.pem:
  x509.private_key_managed:
    - bits: 4096
    - backup: True


/tmp/sample.cert.pem:
  x509.certificate_managed:
    - signing_private_key: /tmp/sample.key.pem
    - CN: localhost
    - days_valid: 300
    - days_remaining: 300
    - managed_private_key:
        name: /tmp/sample.key.pem
        bits: 4096
```

Whether or not it's properly managing the private key, I'm not sure, but that state doesn't fail.

If we remove `managed_private_key` then it *will* fail, with no notice. That's a problem.

Reply:
@waynew it's the `x509.private_key_managed` state that manages the key. Without it the `x509.certificate_managed` state will fail. The `managed_private_key` parameter isn't doing anything. Nobody would write what you've written except by accident. Even if you believed the documentation was correct you wouldn't write that.

If this change just logged a warning and ignored it instead of raising `SaltInvocationError` would that be OK?

Reply:
Ah, good point @waynew, there are definitely examples (albeit mistaken) telling people to do that.

I'd be extremely happy to switch the `SaltInvocationError` to a logged warning, but getting the argument actually working will be a headache. Will that be OK?

Reply:
If we keep the existing behavior, and log a warning telling people that they probably shouldn't use the parameter, since it's useless, I'd be :+1: on that.

We can have a further discussion on whether we should fix or deprecate `managed_private_key` (if we have a different mechanism, it seems silly to have a second, broken way to do it).

Reply:
Thanks, added in https://github.com/saltstack/salt/pull/52935/commits/f120c2257e6f9f1dbb58ea7f788cb5e7d41791e2 with a test case.

Reply:
i'm good with this approach. :) Its a good medium of not breaking states and deprecating in the future. Thanks. Can I just ask that you add the deprecation to the `Aluminium` release so it has a warning in two major releases per the documentation here: https://docs.saltstack.com/en/latest/topics/development/deprecations.html

Reply:
Sorry @Ch3LL I'm not 100% sure what you mean, I'm slow on the uptake 😄 

I read these pages:

https://docs.saltstack.com/en/latest/topics/development/deprecations.html
https://docs.saltstack.com/en/latest/topics/releases/version_numbers.html#code-names

>  For example, if the current release under development is Sodium, the deprecated code and associated warnings should remain in place and warn for at least Aluminum.

This is going in as a 3000.x (Neon) bugfix, right? Neon + 2 versions = Magnesium.

> Can I just ask that you add the deprecation to the Aluminium release

Wait, you don't want any warnings to show until then? `salt.utils.versions.warn_until` will log a warning until the specified version, then it will throw an error, not a deprecation.

Reply:
No worries, i'm sure i was just not clear enough. This would be going into Sodium since its on the head of master, so Sodium + 2 versions would be Aluminium. We recently changed the branching and release strategy: https://github.com/saltstack/salt-enhancement-proposals/pull/20  The idea is we would do less point releases and release more often from master, a goal we are still trying to attain of every 4 months, and hopefully more often when more automation is put in place. A 3000.2 release would not occur unless there are critical enough issues that need to be fixed or CVE release.

I want the warning to show until the Aluminium release, instead of Magnesium. Hopefully that clears it up.

Reply:
To clarify more, Im simply requesting Magnesium be changed to Aluminium instead.

Reply:
Thanks for explaining!

> The idea is we would do less point releases and release more often from master, a goal we are still trying to attain of every 4 months, and hopefully more often when more automation is put in place. A 3000.2 release would not occur unless there are critical enough issues that need to be fixed or CVE release.

Oh great, I didn't realise there were no plans for another 3000.x release. Updated to Aluminium now.